### PR TITLE
[Operator] Add hstack op

### PR DESCRIPTION
**Pull request by zfu82**
### Performance

Perf result on NV-A100
```
benchmark/test_special_perf.py::test_perf_hstack Operator hstack Performance Test (torch.float16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.016384            0.033792
6144                  0.043008            0.033792
11264                   0.0768            0.049152
16384                 0.111616            0.077824
21504                 0.144384             0.09728
26624                 0.176128            0.116736
31744                  0.20992            0.136192
36864                 0.241664            0.154624
41984                 0.275456            0.173056
47104                 0.308224            0.190464
52224                 0.342016            0.212992
57344                 0.374784            0.227328
62464                 0.408576            0.248832
67584                 0.443392            0.268288
72704                 0.475136            0.287744
77824                 0.509952            0.303104
Operator hstack Performance Test (torch.float32)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.013312            0.032768
6144                  0.049152            0.055296
11264                 0.093184            0.100352
16384                  0.13312             0.13824
21504                 0.169984             0.17408
26624                  0.20992            0.211968
31744                 0.247808            0.246784
36864                 0.285696            0.282624
41984                 0.323584            0.323584
47104                 0.362496            0.359424
52224                 0.400384            0.395264
57344                 0.437248            0.429056
62464                  0.47616            0.464896
67584                  0.51712             0.50176
72704                  0.55296            0.539648
77824                 0.589824            0.577536
Operator hstack Performance Test (torch.bfloat16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.014336            0.036864
6144                  0.041984            0.034816
11264                   0.0768            0.049152
16384                 0.111616            0.078848
21504                  0.14336             0.09728
26624                 0.177152             0.11776
31744                 0.208896            0.136192
36864                 0.242688              0.1536
41984                 0.275456             0.17408
47104                 0.309248            0.191488
52224                 0.342016            0.212992
57344                 0.374784            0.228352
62464                 0.408576            0.247808
67584                 0.442368            0.268288
72704                  0.47616            0.287744
77824                 0.509952            0.303104
PASSED
```

### CODE REVIEW

File: src/flag_gems/ops/hstack.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    inp_shapes = [list(_.shape) for _ in tensors]
+    inp0_shape = inp_shapes[0]
+    out_shape = list(inp0_shape)
+    if len(inp0_shape) == 1:
+        dim = 0
+    else:
+        dim = 1
+        for tensor_num, inp_shape in enumerate(inp_shapes[1:]):
+            for i in range(len(inp_shape)):
+                if i != dim and inp_shape[i] != inp0_shape[i]:
```
Comment (iclementine) [inline src/flag_gems/ops/hstack.py]: We should ensure that all tensors have the same ndim, or else inp0_shape[i] may cause an `out of range` error.

torch.hstack would raise an error `Tensors must have same number of dimensions: got {i} and {j}` if any two of the input tensors have the different ndim. with an only exception that tensor of shape `()` would be preprocessed to `(1, )`, (I think this is a not that strict behavior).

If we are going to follow what torch.hsatck does, the following preprocessing or checking should be done beforehand.
1. preprocess 0-d tensors, view them as 1d tensors;
2. check all tensors have the same ndim.

Comment (zfu82) [reply src/flag_gems/ops/hstack.py]: Done

File: tests/test_special_ops.py [hunk 2]
```diff
@@ -532,3 +533,23 @@ def test_accuracy_stack(shape, dim, dtype):
...
+    if dtype in FLOAT_DTYPES:
+        inp = [torch.randn(s, dtype=dtype, device="cuda") for s in shape]
+    else:
+        inp = [
+            torch.randint(low=0, high=0x7FFF, size=s, dtype=dtype, device="cuda").to(
+                dtype
+            )
+            for s in shape
+        ]
+    ref_inp = [to_reference(_, True) for _ in inp]
```
Comment (iclementine) [inline tests/test_special_ops.py]: Since hstack involves no real computation, there is no need to upcast in `to_reference`.

**Review by iclementine — APPROVED** [review]
LGTM

**Pull request merged**